src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,282 @@
+use anyhow::{anyhow, Context, Result};
+extern crate serde;
+use self::serde::{Deserialize, Serialize};
+use super::*;
+use asn1_rs::{oid, Oid, Integer};
+use async_trait::async_trait;
+use openssl::{
+    ec::EcKey,
+    ecdsa,
+    pkey::{PKey, Public},
+    x509,
+};
+use serde_json::json;
+use sev::firmware::guest::types::AttestationReport;
+use sev::firmware::host::types::{CertTable, SnpCertType};
+use sha2::{Digest, Sha384};
+use x509_parser::prelude::*;
+
+#[derive(Serialize, Deserialize)]
+struct SnpEvidence {
+    attestation_report: AttestationReport,
+    cert_chain: CertTable,
+}
+
+const HW_ID_OID: Oid<'static> = oid!(1.3.6 .1 .4 .1 .3704 .1 .4);
+const UCODE_SPL_OID: Oid<'static> = oid!(1.3.6 .1 .4 .1 .3704 .1 .3 .8);
+const SNP_SPL_OID: Oid<'static> = oid!(1.3.6 .1 .4 .1 .3704 .1 .3 .3);
+const TEE_SPL_OID: Oid<'static> = oid!(1.3.6 .1 .4 .1 .3704 .1 .3 .2);
+const LOADER_SPL_OID: Oid<'static> = oid!(1.3.6 .1 .4 .1 .3704 .1 .3 .1);
+
+#[derive(Debug, Default)]
+pub struct Snp {}
+
+#[async_trait]
+impl Verifier for Snp {
+    async fn evaluate(
+        &self,
+        nonce: String,
+        attestation: &Attestation,
+    ) -> Result<TeeEvidenceParsedClaim> {
+        let tee_evidence = serde_json::from_str::<SnpEvidence>(&attestation.tee_evidence)
+            .context("Deserialize Quote failed.")?;
+
+        verify_report_signature(&tee_evidence)?;
+
+        let report = tee_evidence.attestation_report;
+        if report.version != 2 {
+            return Err(anyhow!("Unexpected report version"));
+        }
+
+        if report.vmpl != 0 {
+            return Err(anyhow!("VMPL Check Failed"));
+        }
+
+        let expected_report_data = calculate_expected_report_data(&nonce, &attestation.tee_pubkey);
+        if report.report_data != expected_report_data {
+            return Err(anyhow!("Report Data Mismatch"));
+        }
+
+        parse_tee_evidence(&report)
+    }
+}
+
+fn get_oid<const N: usize>(
+    vcek: &x509_parser::certificate::TbsCertificate,
+    oid: Oid,
+) -> Result<[u8; N]> {
+    let val = vcek
+        .get_extension_unique(&oid)?
+        .ok_or(anyhow!("Oid not found"))?
+        .value;
+
+    val.try_into()
+        .map_err(|_| anyhow!("Oid data has wrong length."))
+}
+
+fn get_oid_int(
+    vcek: &x509_parser::certificate::TbsCertificate,
+    oid: Oid,
+) -> Result<u8> {
+    let val_raw = get_oid::<3>(&vcek, oid)?;
```

Comment:
Do we know that the raw value will always be 3 bytes? What about the ucode, can't that be 4 bytes in size?

Reply:
I think that's why I am getting 3 bytes. Originally I was trying to directly read the bytes, but this gave me the wrong results. Instead I get three bytes and give it to the der parser, which I think will handle this padding thing you mention?

Reply:
This should be fixed now btw.